sam/pileup_test.go

```diff
@@ -253,3 +253,115 @@ func copyPile(p Pile) Pile {
 	}
 	return n
 }
+
+var PileConsensusTests = []struct {
+	p                 Pile
+	c                 Consensus
+	substitutionsOnly bool
+}{
+	{p: Pile{
+		CountF:    [13]int{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+		CountR:    [13]int{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+		InsCountF: map[string]int{"AAT": 2, "AC": 1},
+		InsCountR: map[string]int{"AAT": 1, "AC": 3},
+		DelCountF: map[int]int{6: 16, 5: 2},
+		DelCountR: map[int]int{6: 19, 5: 1},
+	},
+		c: Consensus{
+			Base:      0,
+			Insertion: []dna.Base{},
+			Deletion:  6,
+			Type:      Deletion,
+		},
+		substitutionsOnly: false,
+	},
+	{p: Pile{
+		CountF:    [13]int{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+		CountR:    [13]int{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+		InsCountF: map[string]int{},
+		InsCountR: map[string]int{},
+		DelCountF: map[int]int{},
+		DelCountR: map[int]int{},
+	},
+		c: Consensus{
+			Base:      0,
+			Insertion: []dna.Base{},
+			Deletion:  0,
+			Type:      Undefined,
+		},
+		substitutionsOnly: false,
+	},
+	{p: Pile{
+		CountF:    [13]int{0, 14, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+		CountR:    [13]int{0, 17, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+		InsCountF: map[string]int{"AAT": 2, "AC": 1},
+		InsCountR: map[string]int{"AAT": 1, "AC": 3},
+		DelCountF: map[int]int{6: 2, 5: 2},
+		DelCountR: map[int]int{6: 4, 5: 1},
+	},
+		c: Consensus{
+			Base:      1,
+			Insertion: []dna.Base{},
+			Deletion:  0,
+			Type:      Base,
+		},
+		substitutionsOnly: false,
+	},
+	{p: Pile{
+		CountF:    [13]int{0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+		CountR:    [13]int{0, 4, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+		InsCountF: map[string]int{"AACTT": 42, "AC": 1},
+		InsCountR: map[string]int{"AACTT": 49, "AC": 3},
+		DelCountF: map[int]int{6: 2, 5: 2},
+		DelCountR: map[int]int{6: 4, 5: 1},
+	},
+		c: Consensus{
+			Base:      1,
+			Insertion: dna.StringToBases("AACTT"),
+			Deletion:  0,
+			Type:      Insertion,
+		},
+		substitutionsOnly: false,
+	},
+	{p: Pile{
+		CountF:    [13]int{0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+		CountR:    [13]int{0, 4, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+		InsCountF: map[string]int{"AACTT": 42, "AC": 1},
+		InsCountR: map[string]int{"AACTT": 49, "AC": 3},
+		DelCountF: map[int]int{6: 2, 5: 2},
+		DelCountR: map[int]int{6: 4, 5: 1},
+	},
+		c: Consensus{
+			Base:      1,
+			Insertion: []dna.Base{},
+			Deletion:  0,
+			Type:      Base,
+		},
+		substitutionsOnly: true,
+	},
+}
+
+func TestPileConsensus(t *testing.T) {
```

Comment:
I _**strongly**_ advise testing this code with piles derived with GoPileup from sam records that contain both single base and multi-base deletions and insertions and check that you are getting the expected result. It may even be best to test this by generating a fasta consensus as you are planning. There is a high potential for gotchas here that will need to be carefully tested. For instance, the testing piles above (e.g. newline 326) list 42 bases having an insertion of AACTT on the F strand, but only 2 bases reported in the counts on the forward strand. The only way such a pile would be possible is if you had 40 reads that immediately begin with an insertion, which almost no aligner would allow. Piles are strictly generated per reference base, and since there is no index in the reference fasta for inserted sequence, the insertion is stored in the base immediately prior to the insertion. Therefore when you get an insertion as your consensus, you must also include a consensus for the bases/del in the same pile (i.e. you need 2 consensus structs for every pile where you are calling an insertion). I am disappointed in myself for not noting this in the documentation, but I am almost sure this is the case. If you confirm this to be true during your testing, I would appreciate you noting this in the documentation for the pile struct.